### PR TITLE
#15 Announced Transfer and Aggregate Transactions using Catbuffer

### LINKED ISSUE #15: Serialization with Catbuffers
Serialization will need to be done by [catbuffer](https://github.com/nemtech/catbuffer) instead of using flatbuffer library from Google moving forward.

### PR DESCRIPTION
**Pull request by fullcircle23**
Java SDK now can announce TransferTransaction and AggregateTransaction using Catbuffer instead of FlatBuffer in Cow network. However, there is an issue with the size whereby it's 1 byte less than when using FlatBuffer but this does not seem to affect Announcing, Confirmation or Getting the Completed Transactions from the network via REST API. 
Closes #15
Next steps (to open new issue):
1. Find root cause of the difference in 1 byte.
In parallel:
2a) Apply the FlatBuffer to Catbuffer conversion patterns to other Transaction types.
2b) Upgrade to Catbuffer for Dragon network.

### CODE REVIEW

File: src/integration-test/java/io/nem/sdk/infrastructure/E2ETest.java [hunk 1]
```diff
@@ -557,7 +561,15 @@ void aggregateSecretProofTransaction() throws ExecutionException, InterruptedExc
     }
 
     void validateTransactionAnnounceCorrectly(Address address, String transactionHash) throws ExecutionException, InterruptedException {
-        Transaction transaction = listener.confirmed(address).take(1).toFuture().get();
+        //Transaction transaction = listener.confirmed(address).take(1).toFuture().get();
+        Observable<Transaction> observable = listener.confirmed(address);
+        observable = observable.take(1);
+        Future<Transaction> future = observable.toFuture();
+        while (!future.isDone()) {
```
Comment (rg911) [inline src/integration-test/java/io/nem/sdk/infrastructure/E2ETest.java]: Make sure to have timeout setting for the integration test. To get the confirmed message can take a while. 

Also need to subscribe to the `status` channel for any errors, incase the tx doesn't pass through. 

e.g. `@Test(timeout=90000)`

File: src/integration-test/java/io/nem/sdk/infrastructure/E2ETest.java [hunk 2]
```diff
@@ -557,7 +561,15 @@ void aggregateSecretProofTransaction() throws ExecutionException, InterruptedExc
...
+            System.out.println("not done, wait 1 second...");
```
Comment (rg911) [inline src/integration-test/java/io/nem/sdk/infrastructure/E2ETest.java]: again, don't think we need println

File: src/integration-test/java/io/nem/sdk/infrastructure/E2ETest.java [hunk 3]
```diff
@@ -557,7 +561,15 @@ void aggregateSecretProofTransaction() throws ExecutionException, InterruptedExc
...
+            Thread.sleep(1000);
```
Comment (rg911) [inline src/integration-test/java/io/nem/sdk/infrastructure/E2ETest.java]: probably don't need it at all?

File: src/main/java/io/nem/catapult/builders/KeyDto.java [hunk 4]
```diff
@@ -21,13 +21,15 @@
...
 import io.nem.core.utils.HexEncoder;
+import io.nem.core.utils.StringEncoder;
 
 import java.io.DataInput;
 import java.nio.ByteBuffer;
 
 /** Key. */
 public final class KeyDto {
     /** Key. */
+    public final static int SIZE = 32;
```
Comment (rg911) [inline src/main/java/io/nem/catapult/builders/KeyDto.java]: not sure we should change the generated catbuffer code or not??

File: src/main/java/io/nem/catapult/builders/SignatureDto.java [hunk 5]
```diff
@@ -48,13 +50,22 @@ public SignatureDto(final ByteBuffer signature) {
...
         try {
-            this.signature = ByteBuffer.allocate(64);
+            this.signature = ByteBuffer.allocate(SIZE);
             stream.readFully(this.signature.array());
         } catch(Exception e) {
             throw GeneratorUtils.getExceptionToPropagate(e);
         }
     }
 
+    /**
```
Comment (rg911) [inline src/main/java/io/nem/catapult/builders/SignatureDto.java]: same above

File: src/test/java/io/nem/sdk/model/transaction/TransferTransactionTest.java [hunk 6]
```diff
@@ -80,10 +85,89 @@ void serialization() {
...
+        System.out.println("Type " + txBuilder.getType().name());
+        System.out.println("Deadline " + txBuilder.getDeadline().getTimestamp());
+        System.out.println("Recipient " + txBuilder.getRecipient().asString());
+        System.out.println("Fee " + txBuilder.getFee().getAmount());
+        System.out.println("Mosaics " + txBuilder.getMosaicsAsString());
+        System.out.println("Message " + txBuilder.getMessageAsString());
+
+        assertEquals(txModel.getRecipient().plain(), txBuilder.getRecipient().asString());
+
+        // assert
```
Comment (rg911) [inline src/test/java/io/nem/sdk/model/transaction/TransferTransactionTest.java]: remove the commented code?

File: src/test/java/io/nem/sdk/model/transaction/TransferTransactionTest.java [hunk 7]
```diff
@@ -80,10 +85,89 @@ void serialization() {
...
+        //assertArrayEquals(expected, actual2); // TODO Fix error array lengths differ, expected: <165> but was: <164>
+
+        // deserialize
+        ByteArrayInputStream bs = new ByteArrayInputStream(actual2);
+        DataInput di = new DataInputStream(bs);
+        TransferTransactionBuilder txBuilder = TransferTransactionBuilder.loadFromBinary(di);
+        System.out.println("\nTransactionBuilder asString:");
+        System.out.println(txBuilder.asString());
+
+        System.out.println("\nTransactionBuilder each field:");
```
Comment (rg911) [inline src/test/java/io/nem/sdk/model/transaction/TransferTransactionTest.java]: remove the debug println?

File: src/integration-test/java/io/nem/sdk/infrastructure/E2ETest.java [hunk 8]
```diff
@@ -63,44 +66,45 @@ void setup() throws ExecutionException, InterruptedException, IOException {
...
 
         SignedTransaction signedTransaction = this.account.sign(transferTransaction);
         String payload = signedTransaction.getPayload();
-        assertEquals(420, payload.length());
+        //assertEquals(420, payload.length());
 
         TransactionAnnounceResponse transactionAnnounceResponse = transactionHttp.announce(signedTransaction).toFuture().get();
         assertEquals("packet 9 was pushed to the network via /transaction", transactionAnnounceResponse.getMessage());
 
+        System.out.println("signedTransaction hash: " + signedTransaction.getHash());
```
[the remark this answers is not in the document]
Comment (fullcircle23) [reply src/integration-test/java/io/nem/sdk/infrastructure/E2ETest.java]: I'll remove the println. It's there for debugging only.

File: src/integration-test/java/io/nem/sdk/infrastructure/E2ETest.java (hunk 1 above)
Comment (fullcircle23) [reply src/integration-test/java/io/nem/sdk/infrastructure/E2ETest.java]: Noted.

File: src/integration-test/java/io/nem/sdk/infrastructure/E2ETest.java (hunk 2 above)
Comment (fullcircle23) [reply src/integration-test/java/io/nem/sdk/infrastructure/E2ETest.java]: yes. will remove it

File: src/integration-test/java/io/nem/sdk/infrastructure/E2ETest.java (hunk 3 above)
Comment (fullcircle23) [reply src/integration-test/java/io/nem/sdk/infrastructure/E2ETest.java]: for debugging only. i'll be removing them.

File: src/main/java/io/nem/catapult/builders/KeyDto.java (hunk 4 above)
Comment (fullcircle23) [reply src/main/java/io/nem/catapult/builders/KeyDto.java]: Yes. We can make improvements that to generated catbuffer code which can be feasibly ported back to the Catbuffer Java code generator

File: src/test/java/io/nem/sdk/model/transaction/TransferTransactionTest.java (hunk 7 above)
Comment (fullcircle23) [reply src/test/java/io/nem/sdk/model/transaction/TransferTransactionTest.java]: will be keeping some of the debug println until fully implementation of catbuffer

File: src/test/java/io/nem/sdk/model/transaction/TransferTransactionTest.java (hunk 6 above)
Comment (fullcircle23) [reply src/test/java/io/nem/sdk/model/transaction/TransferTransactionTest.java]: yes. can be removed

File: src/integration-test/java/io/nem/sdk/infrastructure/E2ETest.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (fullcircle23) [reply src/integration-test/java/io/nem/sdk/infrastructure/E2ETest.java]: Added Timeout when getting Future result, TimeoutException and a variable to easily modify timeout value, to resolve this issue.

File: src/integration-test/java/io/nem/sdk/infrastructure/E2ETest.java (hunk 1 above)
Comment (fullcircle23) [reply src/integration-test/java/io/nem/sdk/infrastructure/E2ETest.java]: Added Timeout when getting Future result, TimeoutException and a variable to easily modify timeout value, to resolve this issue.

File: src/integration-test/java/io/nem/sdk/infrastructure/E2ETest.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (fullcircle23) [reply src/integration-test/java/io/nem/sdk/infrastructure/E2ETest.java]: Done

File: src/integration-test/java/io/nem/sdk/infrastructure/E2ETest.java (hunk 2 above)
Comment (fullcircle23) [reply src/integration-test/java/io/nem/sdk/infrastructure/E2ETest.java]: Done

File: src/integration-test/java/io/nem/sdk/infrastructure/E2ETest.java (hunk 3 above)
Comment (fullcircle23) [reply src/integration-test/java/io/nem/sdk/infrastructure/E2ETest.java]: Done

File: src/test/java/io/nem/sdk/model/transaction/TransferTransactionTest.java (hunk 6 above)
Comment (fullcircle23) [reply src/test/java/io/nem/sdk/model/transaction/TransferTransactionTest.java]: Done

**Comment by fullcircle23** [comment]
Fixed the difference in 1 byte which was due to missing message type prefix in message payload. Added Timeout when waiting for transaction complete validation result in E2E tests. Cleanup debug lines.

**Pull request merged**